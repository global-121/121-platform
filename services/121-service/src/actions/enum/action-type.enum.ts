export enum ActionType {
  startBlockNewPayment = 'start-block-new-payment',
  endBlockNewPayment = 'end-block-new-payment',
}
