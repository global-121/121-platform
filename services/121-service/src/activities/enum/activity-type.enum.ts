export enum ActivityTypeEnum {
  DataChange = 'data-change',
  FinancialServiceProviderChange = 'financial-service-provider-change',
  Message = 'message',
  Note = 'note',
  StatusChange = 'status-change',
  Transaction = 'transaction',
}
