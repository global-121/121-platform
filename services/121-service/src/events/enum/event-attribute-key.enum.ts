export enum EventAttributeKeyEnum {
  reason = 'reason',
  oldValue = 'oldValue',
  newValue = 'newValue',
  fieldName = 'fieldName',
  duplicateWithRegistrationId = 'duplicateWithRegistrationId',
  duplicateWithRegistrationProgramId = 'duplicateWithRegistrationProgramId',
}
