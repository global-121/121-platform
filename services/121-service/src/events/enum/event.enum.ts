export enum EventEnum {
  registrationDataChange = 'registrationDataChange',
  financialServiceProviderChange = 'financialServiceProviderChange',
  registrationStatusChange = 'registrationStatusChange',
  ignoredDuplicate = 'ignoredDuplicate',
}
