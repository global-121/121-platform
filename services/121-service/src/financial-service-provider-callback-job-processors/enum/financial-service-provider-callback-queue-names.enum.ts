export enum FinancialServiceProviderCallbackQueueNames {
  safaricomTransferCallback = 'safaricomTransferCallback',
  safaricomTimeoutCallback = 'safaricomTimeoutCallback',
}
