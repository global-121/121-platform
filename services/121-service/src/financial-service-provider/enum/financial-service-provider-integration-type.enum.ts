export enum FinancialServiceProviderIntegrationType {
  api = 'api',
  csv = 'csv',
  xml = 'xml',
}
