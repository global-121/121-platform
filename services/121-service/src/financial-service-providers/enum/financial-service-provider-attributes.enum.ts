export enum FinancialServiceProviderAttributes {
  phoneNumber = 'phoneNumber',
  nationalId = 'nationalId',
  fullName = 'fullName',
  addressStreet = 'addressStreet',
  addressHouseNumber = 'addressHouseNumber',
  addressHouseNumberAddition = 'addressHouseNumberAddition',
  addressPostalCode = 'addressPostalCode',
  addressCity = 'addressCity',
  bankAccountNumber = 'bankAccountNumber',
  whatsappPhoneNumber = 'whatsappPhoneNumber',
}
