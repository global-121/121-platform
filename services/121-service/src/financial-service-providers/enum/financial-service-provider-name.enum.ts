export enum FinancialServiceProviderName {
  intersolveVoucherWhatsapp = 'Intersolve-voucher-whatsapp',
  intersolveVoucherPaper = 'Intersolve-voucher-paper',
  intersolveVisa = 'Intersolve-visa',
  intersolveJumboPhysical = 'Intersolve-jumbo-physical',
  africasTalking = 'Africas-talking',
  belcash = 'BelCash',
  vodacash = 'VodaCash',
  bobFinance = 'BoB-finance',
  ukrPoshta = 'UkrPoshta',
  safaricom = 'Safaricom',
  commercialBankEthiopia = 'Commercial-bank-ethiopia',
  excel = 'Excel',
}

export enum FinancialServiceProviderConfigurationEnum {
  password = 'password',
  username = 'username',
  columnsToExport = 'columnsToExport',
  columnToMatch = 'columnToMatch',
  brandCode = 'brandCode',
  displayName = 'displayName',
  coverLetterCode = 'coverLetterCode',
}

export const FinancialServiceProviderConfigurationMapping: {
  [key in FinancialServiceProviderName]?: any;
} = {
  [FinancialServiceProviderName.intersolveVoucherWhatsapp]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVoucherPaper]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVisa]: [
    FinancialServiceProviderConfigurationEnum.brandCode,
    FinancialServiceProviderConfigurationEnum.coverLetterCode,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveJumboPhysical]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.africasTalking]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.belcash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.vodacash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.bobFinance]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.ukrPoshta]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.safaricom]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.commercialBankEthiopia]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.excel]: [
    FinancialServiceProviderConfigurationEnum.columnsToExport,
    FinancialServiceProviderConfigurationEnum.columnToMatch,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
};
