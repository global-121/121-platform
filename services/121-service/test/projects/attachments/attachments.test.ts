import { HttpStatus } from '@nestjs/common';

import { SeedScript } from '@121-service/src/scripts/enum/seed-script.enum';
import {
  deleteAttachment,
  getAttachment,
  getAttachments,
  uploadAttachment,
} from '@121-service/test/helpers/project-attachments.helper';
import {
  getAccessToken,
  resetDB,
} from '@121-service/test/helpers/utility.helper';
import { projectIdPV } from '@121-service/test/registrations/pagination/pagination-data';

const testImagePath = './test-attachment-data/sample.jpg';
const testImageFilename = 'Test Image';
const testImageMimetype = 'image/jpeg';
const testImageExtension = 'jpg';

const testCsvPath = './test-registration-data/test-registrations-OCW.csv';

describe('Project Attachments', () => {
  let accessToken: string;

  beforeEach(async () => {
    await resetDB(SeedScript.nlrcMultiple, __filename);
    accessToken = await getAccessToken();
  });

  it('should upload a document attachment to a project', async () => {
    // Arrange

    // Act
    const response = await uploadAttachment({
      projectId: projectIdPV,
      filePath: testImagePath,
      filename: testImageFilename,
      accessToken,
    });

    // Assert
    expect(response.status).toBe(HttpStatus.CREATED);
    expect(response.body).toHaveProperty('id');
  });

  it('should list all attachments for a project', async () => {
    // Arrange - Upload a test attachment first
    await uploadAttachment({
      projectId: projectIdPV,
      filePath: testImagePath,
      filename: testImageFilename,
      accessToken,
    });

    // Act
    const response = await getAttachments({
      projectId: projectIdPV,
      accessToken,
    });

    // Assert
    expect(response.status).toBe(HttpStatus.OK);
    expect(Array.isArray(response.body)).toBe(true);
    expect(response.body).toHaveLength(1); // Should return the uploaded attachment

    const attachment = response.body[0];
    expect(attachment).toMatchSnapshot({
      created: expect.any(String),
      updated: expect.any(String),
    });
  });

  it('should download a specific attachment', async () => {
    // Arrange - Upload a test attachment first
    const uploadResponse = await uploadAttachment({
      projectId: projectIdPV,
      filePath: testImagePath,
      filename: testImageFilename,
      accessToken,
    });

    const attachmentId = uploadResponse.body.id;

    // Act
    const response = await getAttachment({
      projectId: projectIdPV,
      attachmentId,
      accessToken,
    })
      .buffer()
      .parse((res, callback) => {
        // Collect the response data as a buffer
        const chunks: Buffer[] = [];
        res.on('data', (chunk) => {
          chunks.push(chunk);
        });
        res.on('end', () => {
          callback(null, Buffer.concat(chunks));
        });
      });

    // Assert
    expect(response.status).toBe(HttpStatus.OK);
    expect(response.headers['content-type']).toBe(testImageMimetype);
    expect(response.headers['content-disposition']).toContain(
      `attachment; filename="${testImageFilename}.${testImageExtension}"`,
    );
    expect(response.body.length).toBeGreaterThan(0);
  });

  it('should delete a specific attachment', async () => {
    const uploadResponse = await uploadAttachment({
      projectId: projectIdPV,
      filePath: testImagePath,
      filename: testImageFilename,
      accessToken,
    });

    const attachmentId = uploadResponse.body.id;

    // Act
    const deleteResponse = await deleteAttachment({
      projectId: projectIdPV,
      attachmentId,
      accessToken,
    });

    // Assert - Deletion successful
    expect(deleteResponse.status).toBe(HttpStatus.NO_CONTENT);

    // Verify the attachment is no longer retrievable
    const getResponse = await getAttachment({
      projectId: projectIdPV,
      attachmentId,
      accessToken,
    });

    expect(getResponse.status).toBe(HttpStatus.NOT_FOUND);
  });

  it('should reject files with invalid mime types', async () => {
    // Act
    const response = await uploadAttachment({
      projectId: projectIdPV,
      filePath: testCsvPath,
      filename: 'Test CSV',
      accessToken,
    });

    // Assert
    expect(response.status).toBe(HttpStatus.BAD_REQUEST);
  });

  it('should handle attachment not found', async () => {
    // Act
    const response = await getAttachment({
      projectId: projectIdPV,
      attachmentId: 999999,
      accessToken,
    });

    // Assert
    expect(response.status).toBe(HttpStatus.NOT_FOUND);
  });
});
