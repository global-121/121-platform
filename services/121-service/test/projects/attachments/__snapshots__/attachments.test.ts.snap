// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Attachments should list all attachments for a project 1`] = `
{
  "created": Any<String>,
  "filename": "Test Image.jpg",
  "id": 1,
  "mimetype": "image/jpeg",
  "projectId": 2,
  "updated": Any<String>,
  "user": {
    "id": 1,
    "username": "admin@example.org",
  },
}
`;
